book/src/04_4_1_interrupts.md

```diff
@@ -12,77 +11,25 @@ cargo run --example solution
-
-    ```rust
-    static mut EVENT_QUEUE: Option<QueueHandle_t> = None;
-    ```
-
-5. Create the event queue using [`pub unsafe extern "C" fn xQueueGenericCreate`](https://esp-rs.github.io/esp-idf-sys/esp_idf_sys/fn.xQueueGenericCreate.html). This lets us safely pass events from an interrupt routine to our main thread.
-
-    ```rust
-    EVENT_QUEUE = Some(xQueueGenericCreate(QUEUE_SIZE, ITEM_SIZE, QUEUE_TYPE_BASE));
-    ```
-
-6. Add a function which that will be called whenever there is a GPIO interrupt on our button pin. We put this function in a special block of RAM (`iram0`), so it will still be available even if the external flash is busy doing something else (like filesystem work). The function needs to get the queue handle from `EVENT_QUEUE` and call the `xQueueGiveFromISR` function with a `std::ptr::null_mut()` - the objects in our queue are of size zero, so we don't actually need a 'thing' to put on the queue. Instead, the act of pushing a 'nothing' is enough to wake up the other end!
-
-    ```rust
-    #[link_section = ".iram0.text"]
-    unsafe extern "C" fn button_interrupt(_: *mut c_void) {
-        xQueueGiveFromISR(EVENT_QUEUE.unwrap(), std::ptr::null_mut());
-    }
-    ```
-
-If the interrupt fires, an event is added to the queue.
-
-7. Pass the function we just wrote to the generic GPIO interrupt handler we registered earlier, along with the number of the GPIO pin that should cause this function to be executed.
-
-    ```rust
-    esp!(gpio_isr_handler_add(
-        GPIO_NUM,
-        Some(button_interrupt),
-        std::ptr::null_mut()
-    ))?;
-    ```
-
-8. Inside a loop, wait until the queue has an item in it. That is, until the `button_interrupt` function puts something in the queue.
-
-    ```rust
-    let res = xQueueReceive(EVENT_QUEUE.unwrap(), ptr::null_mut(), QUEUE_WAIT_TICKS);
-    ```
-
-9. Handle the value of `res`, so that "Button pushed!" is logged, if the button is pushed.
-
-10. Run the program and push the `BOOT` button, so see how it works!
+2. Instantiate a new notification and notifier
+    - See `hal::task::notification` documentation
+3. In an `unsafe` block, create a subscription and its callback function.
+    - See `PinDriver::subscribe` and `task::notify_and_yield`
+    - The reasons for being `unsafe` are:
+      - The callback function will run in the [ISR (Interrupt Service Routine)](https://en.wikipedia.org/wiki/Interrupt_handler), so we should avoid calling any blocking functions on it, this includes STD, `libc` or FreeRTOS APIs (except for a few allowed ones).
+      - Callback closure is capturing its environment and you can use static variables inserted onto it. Captured variables need to outlive the subscription. You can also, use non-static variables, but that requires extra caution, see `esp_idf_hal::gpio::PinDriver::subscribe_nonstatic` documentation for more details.
```

Comment:
Just a small nit for explaining stuff: If the user uses a closure here than this is true(most people will do) , people might read the argument of the subscribe method. If there are some guys that implement FnMut themself and pass it here it must not be true :D .. I just say it because for new guys to rust  the argument of the subscribe method : `F: FnMut() + Send + 'static` is maybe absolutely not clear. Do we use callback before this point inside the training?

Reply:
> Do we use callback before this point inside the training?

We have a [closure for receiving MQTT messages](https://esp-rs.github.io/std-training/03_5_3_mqtt.html#handling-incoming-messages)